# Strategy 2: Map sequencing reads (BAM/fragments.tsv/.bed) to cPeaks with Map2cpeak

in map2cpeaks folder, download it and use it, you can try to run a dome in the dome folder

## Version
Python (>=3.7)

## Requirments

```
numpy
gzip
tqdm
```

## Method 1: Map the sequencing reads (fragments.tsv.gz) in each sample/cell to generate cell-by-cPeak matrix (.mtx/.h5ad)

```bash
usage:

1.


cd map2cpeaks


2. 

python main.py -f path/to/your_fragment.tsv.gz
               
--fragment_path, -f: the input file must be *.tsv.gz file. If barcode_path is None, tsv file need contain 4 columns:chr, start, end, barcode, sep with '\t'. If barcode_path is  not None, tsv file need contain 3 columns:chr, start, end, sep with '\t';
 
optional arguments:

 --help, -h:          show this help message
 --barcode_path, -b:  Each line is a barcode, the code will use the barcodes in the file, Default to use all barcodes in fragment
 --output, -o:        output folder, Default to ./map2cpeaks_result
 --output_name:       name of output files, Default to cell-cpeaks.
 --num_cores, -n:     number of cores to use, Default to 10.
 --reference:         cPeak version, hg38 or hg19, Default to hg38.
```

The output file contains a barcode.txt and an mtx file that stores the matrix of map results.

## Method 2. Directly map the pre-identified features like peaks to cPeaks (NOT recommand)

**This is not a good idea.** It may lose information in the genomic regions which are not included in pre-identfied features. Also, for bulk ATAC-seq data, the quantification of each cPeak is inaccurate.

```bash
usage: python main.py [--bed_path feature.bed]

--bed_path, -bed: the input feature.bed file, for example, MACS2calledPeaks.bed.
```