{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "629c1b78-125c-4a62-84b0-d4658f7c58ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########################################################\n",
    "# Matt Regner\n",
    "# Franco Lab\n",
    "# Date: May-December 2020\n",
    "# \n",
    "\n",
    "# Description: This script performs the following tasks  \n",
    "#         1) scRNA-seq processing before doublet removal\n",
    "#         2) Doublet detection and removal\n",
    "#         3) scRNA-seq processing after doublet removal \n",
    "#         4) SingleR cell typing \n",
    "###########################################################\n",
    "\n",
    "# Change to your working directory \n",
    "dir <- \".\"\n",
    "setwd(dir) \n",
    "\n",
    "#Change to your R library path\n",
    "########################################################################\n",
    "#.libPaths('/home/regnerm/anaconda3/envs/r-environment/lib/R/library')\n",
    "source(\"./rowr.R\")\n",
    "########################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ad9274ab-a0b8-4670-ba3c-18aef9800df1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#library(scater)\n",
    "library(dplyr)\n",
    "library(Seurat)\n",
    "library(patchwork)\n",
    "library(SingleCellExperiment)\n",
    "#library(ComplexHeatmap)\n",
    "#library(ConsensusClusterPlus)\n",
    "#library(msigdbr)\n",
    "#library(fgsea)\n",
    "library(dplyr)\n",
    "library(tibble)\n",
    "#library(DoubletFinder)\n",
    "library(Signac)\n",
    "library(ggplot2)\n",
    "library(stringr)\n",
    "#library(SingleR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e5b1d38-c462-42dc-9c0f-03104d5ed719",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define filepaths/variables/signature sets:\n",
    "###############################################\n",
    "\n",
    "# PanglaoDB \n",
    "tsv=gzfile(\"./PanglaoDB_markers_27_Mar_2020.tsv.gz\")  \n",
    "panglaodb <- read.csv(tsv,header=T,sep = \"\\t\") \n",
    "panglaodb <- dplyr::filter(panglaodb,species == \"Hs\" | species == \"Mm Hs\")# Human subset \n",
    "panglaodb <- dplyr::filter(panglaodb,organ == \"Connective tissue\" |\n",
    "                             organ == \"Epithelium\" |\n",
    "                             organ == \"Immune system\" |\n",
    "                             organ == \"Reproductive\"|\n",
    "                             organ == \"Vasculature\" |\n",
    "                             organ == \"Smooth muscle\"\n",
    ")\n",
    "panglaodb <- split(as.character(panglaodb$official.gene.symbol), panglaodb$cell.type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cf51152c-2b11-4a92-94b5-01c24a60414b",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in Read10X_h5(filename = \"./filtered_feature_bc_matrix.h5\"): Please install hdf5r to read HDF5 files\n",
     "output_type": "error",
     "traceback": [
      "Error in Read10X_h5(filename = \"./filtered_feature_bc_matrix.h5\"): Please install hdf5r to read HDF5 files\nTraceback:\n",
      "1. Read10X_h5(filename = \"./filtered_feature_bc_matrix.h5\")",
      "2. stop(\"Please install hdf5r to read HDF5 files\")"
     ]
    }
   ],
   "source": [
    "# ESTIMATE signatures \n",
    "ESTIMATE.signatures <- \"./ESTIMATE_signatures.csv\"\n",
    "\n",
    "GRCH38.annotations <- \"./Homo_sapiens.GRCh38.86.txt\"\n",
    "\n",
    "doublet.rate = 0.0460\n",
    "\n",
    "SAMPLE.ID = \"endo_3533EL\"\n",
    "\n",
    "###########################################################\n",
    "# Part 1: scRNA-seq processing before doublet removal\n",
    "###########################################################\n",
    "\n",
    "# Load the RNA dataset\n",
    "counts <- Read10X_h5(filename = \"./filtered_feature_bc_matrix.h5\")\n",
    "# Initialize the Seurat object with the raw (non-normalized data).\n",
    "rna <- CreateSeuratObject(counts = counts, min.cells = 3)# genes not present in at least 3 cells are removed \n",
    "rna\n",
    "\n",
    "PreQCNumCells <- length(colnames(rna))\n",
    "\n",
    "# The [[ operator can add columns to object metadata. This is a great place to stash QC stats\n",
    "rna[[\"percent.mt\"]] <- PercentageFeatureSet(rna, pattern = \"^MT-\")\n",
    "\n",
    "\n",
    "# QC metrics: nCount_RNA, nFeature_RNA, and percent mitochrondial counts\n",
    "# Outliers are >2 MADs\n",
    "rna@meta.data$nCount_RNA_outlier_2mad <- isOutlier(log(rna@meta.data$nCount_RNA),log = F,type = \"lower\",nmads = 2)\n",
    "rna@meta.data$nFeature_RNA_outlier_2mad <- isOutlier(log(rna@meta.data$nFeature_RNA),log = F,type = \"lower\",nmads = 2)\n",
    "rna@meta.data$percent_mt_outlier_2mad <- isOutlier(log1p(rna@meta.data$percent.mt),log = F,type = \"higher\",nmads = 2)\n",
    "\n",
    "rna <- subset(rna, subset = nCount_RNA_outlier_2mad == \"FALSE\" &\n",
    "                nFeature_RNA_outlier_2mad == 'FALSE' & \n",
    "                percent_mt_outlier_2mad == \"FALSE\")\n",
    "\n",
    "PostQCNumCells <- length(colnames(rna))\n",
    "\n",
    "# Default Seurat processing\n",
    "rna <- NormalizeData(rna, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
    "\n",
    "#Feature Selection\n",
    "rna <- FindVariableFeatures(rna, selection.method = \"vst\", nfeatures = 2000)\n",
    "\n",
    "#Scaling & PCA\n",
    "all.genes <- rownames(rna)\n",
    "rna <- ScaleData(rna, features = all.genes)\n",
    "rna <- RunPCA(rna)\n",
    "\n",
    "\n",
    "# Score cells for immune/stromal/fibroblast/endothelial signatures\n",
    "############################################################################################\n",
    "\n",
    "immune.stromal <- read.csv(ESTIMATE.signatures,header = F)\n",
    "\n",
    "stromal <- immune.stromal$V1[1:141]\n",
    "immune <- immune.stromal$V1[142:282]\n",
    "fibroblast <- panglaodb$Fibroblasts\n",
    "endothelial <- panglaodb$`Endothelial cells`\n",
    "epithelial <- panglaodb$`Epithelial cells`\n",
    "smooth <- panglaodb$`Smooth muscle cells`\n",
    "plasma <- panglaodb$`Plasma cells`\n",
    "\n",
    "feats <- list(stromal,immune,fibroblast,endothelial,epithelial,smooth,plasma)\n",
    "\n",
    "rna <- AddModuleScore(rna,features = feats,name = c(\"stromal.\",\"immune.\",\"fibroblast.\",\"endothelial.\",\n",
    "                                                    \"epithelial.\",\"smooth.\",\"plasma.\"),search = T)\n",
    "#########################################################################\n",
    "\n",
    "#######################################################################\n",
    "\n",
    "# Add PC1 to metadata\n",
    "rna@meta.data$PC1 <- rna@reductions$pca@cell.embeddings[,1]\n",
    "\n",
    "count_cor_PC1 <- cor(rna$PC1,rna$nCount_RNA,method = \"spearman\")\n",
    "\n",
    "stromal.cor <- cor(rna$PC1,rna$stromal.1,method = \"spearman\")\n",
    "immune.cor <- cor(rna$PC1,rna$immune.2,method = \"spearman\")\n",
    "fibroblast.cor <- cor(rna$PC1,rna$fibroblast.3,method = \"spearman\")\n",
    "endothelial.cor <- cor(rna$PC1,rna$endothelial.4,method = \"spearman\")\n",
    "epithelial.cor <- cor(rna$PC1,rna$epithelial.5,method = \"spearman\")\n",
    "smooth.cor <- cor(rna$PC1,rna$smooth.6,method = \"spearman\")\n",
    "plasma.cor <- cor(rna$PC1,rna$plasma.7,method = \"spearman\")\n",
    "\n",
    "\n",
    "# Make JackStraw Plot\n",
    "rna <- JackStraw(rna, num.replicate = 100,dims = 50)\n",
    "rna <- ScoreJackStraw(rna,dims = 1:50)\n",
    "JackStrawPlot(rna, dims = 1:50)+ggsave(\"JackStraw_predoublet.png\")\n",
    "# If PC1 is correalted with read depth, check to see if biological variation is corralted to PC1\n",
    "if (round(abs(count_cor_PC1),2) > 0.5){\n",
    "  \n",
    "  if( round(abs(stromal.cor),2) >= 0.5 |\n",
    "      round(abs(immune.cor),2) >= 0.5 |\n",
    "      round(abs(fibroblast.cor),2) >= 0.5 |\n",
    "      round(abs(endothelial.cor),2) >= 0.5 |\n",
    "      round(abs(epithelial.cor),2) >= 0.5 |\n",
    "      round(abs(smooth.cor),2) >= 0.5 |\n",
    "      round(abs(plasma.cor),2) >= 0.5){\n",
    "    rna <- FindNeighbors(rna,dims = 1:50)\n",
    "    rna <- FindClusters(rna,resolution = 0.7)\n",
    "    rna <- RunUMAP(rna,dims = 1:50)\n",
    "    Idents(rna) <- \"RNA_snn_res.0.7\"\n",
    "    \n",
    "    # Verify with inferCNV: is PC1 correlated with CNV events/Malignancy?\n",
    "    #########################################################################\n",
    "    # inferCNV: does PC1 also correlated with CNV/malignancy status?\n",
    "    library(infercnv)\n",
    "    library(stringr)\n",
    "    library(Seurat)\n",
    "    counts_matrix = GetAssayData(rna, slot=\"counts\")\n",
    "    \n",
    "    # Identify immune clusters\n",
    "    #######################################################\n",
    "    # Find immune cells by relative enrichment of ESTIMATE immune signature\n",
    "    library(psych)\n",
    "    test <- VlnPlot(rna,features = \"immune.2\")\n",
    "    data <- describeBy(test$data$immune.2, test$data$ident, mat = TRUE)\n",
    "    data.immune <- dplyr::filter(data,median > 0.1)\n",
    "    \n",
    "    test <- VlnPlot(rna,features = \"plasma.7\")\n",
    "    data <- describeBy(test$data$plasma.7, test$data$ident, mat = TRUE)\n",
    "    data.plasma <- dplyr::filter(data,median > 0.1)\n",
    "    \n",
    "    immune.clusters <- intersect(data.immune$group1,levels(Idents(rna)))\n",
    "    plasma.clusters <- intersect(data.plasma$group1,levels(Idents(rna)))\n",
    "    \n",
    "    immune.clusters <- unique(append(immune.clusters,plasma.clusters))\n",
    "    \n",
    "    for (i in 1:length(immune.clusters)){\n",
    "      j <- which(levels(Idents(rna)) == immune.clusters[i])\n",
    "      levels(Idents(rna))[j] <- paste0(\"immune.\",immune.clusters[i])\n",
    "    }\n",
    "    rna@meta.data$predoublet.idents <- Idents(rna)\n",
    "    idents <- data.frame(rownames(rna@meta.data),rna@meta.data$predoublet.idents)\n",
    "    \n",
    "    colnames(idents) <- c(\"V1\",\"V2\")\n",
    "    \n",
    "    saveRDS(rna,\"./rna_predoublet_preinferCNV.rds\")\n",
    "    # Make inferCNV input files\n",
    "    rownames(idents) <- NULL\n",
    "    colnames(idents) <- NULL\n",
    "    write.table(idents,\"./sample_annotation_file_inferCNV.txt\",sep = \"\\t\",row.names = FALSE)\n",
    "    idents <- read.delim(\"./sample_annotation_file_inferCNV.txt\",header = F)\n",
    "    \n",
    "    \n",
    "    gtf <- read.delim(GRCH38.annotations,header = F)\n",
    "    library(EnsDb.Hsapiens.v86)\n",
    "    convert.symbol = function(Data){\n",
    "      ensembls <- Data$V1\n",
    "      ensembls <- gsub(\"\\\\.[0-9]*$\", \"\", ensembls)\n",
    "      geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v86, keys= ensembls, keytype = \"GENEID\", columns = \"SYMBOL\")\n",
    "      Data <- cbind.fill(Data, geneIDs1, fill = NA)\n",
    "      Data <- na.omit(Data)\n",
    "      Data$feature <- Data$SYMBOL\n",
    "      Data.new <- data.frame(Data$SYMBOL,Data$V2,Data$V3,Data$V4)\n",
    "      Data.new$Data.V2 <- paste(\"chr\",Data.new$Data.V2,sep = \"\")\n",
    "      Data.new$Data.SYMBOL <- make.unique(Data.new$Data.SYMBOL)\n",
    "      return(Data.new)\n",
    "    }\n",
    "    gtf <- convert.symbol(gtf)\n",
    "    head(gtf)\n",
    "    \n",
    "    write.table(gtf,\"./Homo_sapiens.GRCh38.86.symbol.txt\",sep = \"\\t\",row.names = FALSE,col.names = FALSE)\n",
    "    \n",
    "    \n",
    "    num.immune.clusters = length(immune.clusters)\n",
    "    # create the infercnv object\n",
    "    if ( num.immune.clusters == 1) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=paste0(\"immune.\",immune.clusters[1]))\n",
    "      \n",
    "    } else if (num.immune.clusters == 2) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2])))\n",
    "      \n",
    "    } else if ( num.immune.clusters == 3 ) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3])))\n",
    "    } else if (num.immune.clusters == 4) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4])))\n",
    "    } else if (num.immune.clusters == 5) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5])))\n",
    "    } else if (num.immune.clusters == 6) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6])))\n",
    "    }else if (num.immune.clusters == 7) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7])))\n",
    "    }else if (num.immune.clusters == 8) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8])))\n",
    "    }else if (num.immune.clusters == 9) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[9])))\n",
    "    }else if (num.immune.clusters == 10) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[9]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[10])))\n",
    "    }\n",
    "    \n",
    "    # perform infercnv operations to reveal cnv signal\n",
    "    infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                                 cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                                 out_dir=\"./output_dir_CNV_predoublet\",  # dir is auto-created for storing outputs\n",
    "                                 cluster_by_groups=T,   # cluster\n",
    "                                 denoise=T,scale_data = T,\n",
    "                                 HMM=T,HMM_type = \"i6\",analysis_mode = \"samples\",min_cells_per_gene = 10,\n",
    "                                 BayesMaxPNormal = 0.4, num_threads = 8\n",
    "                                 \n",
    "    )\n",
    "    \n",
    "    \n",
    "    \n",
    "    regions <- read.delim(\"./output_dir_CNV_predoublet/HMM_CNV_predictions.HMMi6.hmm_mode-samples.Pnorm_0.4.pred_cnv_regions.dat\")\n",
    "    probs <- read.delim(\"./output_dir_CNV_predoublet/BayesNetOutput.HMMi6.hmm_mode-samples/CNV_State_Probabilities.dat\")\n",
    "    \n",
    "    probs <- as.data.frame(t(probs[3,]))\n",
    "    colnames(probs) <- \"Prob.Normal\"\n",
    "    probs <- dplyr::filter(probs,Prob.Normal < 0.05)\n",
    "    cnvs <- rownames(probs)\n",
    "    cnvs <- gsub(\"\\\\.\",\"-\",cnvs)\n",
    "    \n",
    "    regions <- regions[regions$cnv_name %in% cnvs, ]\n",
    "    \n",
    "    \n",
    "    cnv.groups <- sub(\"\\\\..*\", \"\", regions$cell_group_name)\n",
    "    \n",
    "    \n",
    "    length(which(rownames(rna@reductions$pca@cell.embeddings) == rownames(rna@meta.data)))\n",
    "    rna$PC1.loading <- rna@reductions$pca@cell.embeddings[,1]\n",
    "    rna$cell.barcode <- rownames(rna@meta.data)\n",
    "    rna$CNV.Pos <- ifelse(as.character(rna$predoublet.idents) %in% cnv.groups,1,0)\n",
    "    \n",
    "    \n",
    "    cnv.freq <- data.frame(table(regions$cell_group_name))\n",
    "    cnv.freq$Var1 <- sub(\"\\\\..*\", \"\", cnv.freq$Var1)\n",
    "    \n",
    "    rna$Total_CNVs <- ifelse(as.character(rna$predoublet.idents) %in% cnv.freq$Var1,cnv.freq$Freq,0)\n",
    "\n",
    "    boxplot.cnv <- ggplot(rna@meta.data,aes(x= predoublet.idents,y=PC1.loading,color = as.factor(CNV.Pos)))+geom_boxplot()\n",
    "    boxplot.cnv+ggsave(\"Predoublet_CNV_PC1_boxplot.png\")\n",
    "    \n",
    "    data <- describeBy(boxplot.cnv$data$PC1.loading, boxplot.cnv$data$predoublet.idents, mat = TRUE)\n",
    "    data$CNV <- ifelse(data$group1 %in% cnv.groups,1,0)\n",
    "    \n",
    "    wilcox <- wilcox.test(data = rna@meta.data,PC1.loading~CNV.Pos)\n",
    "    \n",
    "    if (wilcox$p.value < 0.05){\n",
    "      rna <- rna\n",
    "      library(stringr)\n",
    "      levels(Idents(rna)) <- str_remove(levels(Idents(rna)),\"immune.\")\n",
    "      saveRDS(rna,\"./rna_predoublet_PassedPC1Checks.rds\")\n",
    "    }else{\n",
    "      all.genes <- rownames(rna)\n",
    "      rna <- ScaleData(rna, features = all.genes,vars.to.regress = \"nCount_RNA\")\n",
    "      rna <- FindNeighbors(rna,dims = 1:50)\n",
    "      rna <- FindClusters(rna,resolution = 0.7)\n",
    "      rna <- RunUMAP(rna,dims = 1:50)\n",
    "      Idents(rna) <- \"RNA_snn_res.0.7\"\n",
    "      \n",
    "      saveRDS(rna,\"./rna_predoublet_FailedCNVTest.rds\")\n",
    "    }\n",
    "\n",
    "  }else{\n",
    "    all.genes <- rownames(rna)\n",
    "    rna <- ScaleData(rna, features = all.genes,vars.to.regress = \"nCount_RNA\")\n",
    "    rna <- FindNeighbors(rna,dims = 1:50)\n",
    "    rna <- FindClusters(rna,resolution = 0.7)\n",
    "    rna <- RunUMAP(rna,dims = 1:50)\n",
    "    Idents(rna) <- \"RNA_snn_res.0.7\"\n",
    "    saveRDS(rna,\"./rna_predoublet_FailedCorTest.rds\")\n",
    "  }\n",
    "}else{\n",
    "  rna <- FindNeighbors(rna,dims = 1:50)\n",
    "  rna <- FindClusters(rna,resolution = 0.7)\n",
    "  rna <- RunUMAP(rna,dims = 1:50)\n",
    "  Idents(rna) <- \"RNA_snn_res.0.7\"\n",
    "  saveRDS(rna,\"./rna_predoublet_SkipChecks.rds\")\n",
    "}\n",
    "\n",
    "###########################################################\n",
    "# Part 2: Doublet detection and removal\n",
    "###########################################################\n",
    "\n",
    "#' # Doublet detection: 1) DoubletDecon 2) DoubletFinder 3) Take intersection of calls\n",
    "######################################################################################\n",
    "\n",
    "# # Doublet Decon\n",
    "# # Add if else statement to regress out nCount RNA if needed before running DoubletDecon\n",
    "library(DoubletDecon)\n",
    "idents.length <- length(levels(Idents(rna)))\n",
    "for (i in levels(Idents(rna))){\n",
    "  i <- as.numeric(i)\n",
    "  levels(Idents(rna))[i] <- i -1\n",
    "}\n",
    "\n",
    "#Improved_Seurat_Pre_Process()\n",
    "#Idents(rna) <- as.factor(Idents(rna))\n",
    "seuratObject=rna\n",
    "newFiles=Improved_Seurat_Pre_Process(seuratObject, num_genes=50, write_files=F)\n",
    "\n",
    "#dev.off()\n",
    "results=Main_Doublet_Decon(rawDataFile=newFiles$newExpressionFile,\n",
    "                           groupsFile=newFiles$newGroupsFile,\n",
    "                           filename=SAMPLE.ID,\n",
    "                           location=\"./DoubletDecon\",\n",
    "                           fullDataFile=NULL,\n",
    "                           removeCC=FALSE,\n",
    "                           species=\"hsa\",\n",
    "                           rhop=1.1,\n",
    "                           write=F,\n",
    "                           PMF=TRUE,\n",
    "                           useFull=FALSE,\n",
    "                           heatmap=FALSE,\n",
    "                           centroids=TRUE,\n",
    "                           num_doubs=100,\n",
    "                           only50=FALSE,\n",
    "                           min_uniq=4,\n",
    "                           nCores=4)\n",
    "\n",
    "decon.doublets <- rownames(results$Final_doublets_groups)\n",
    "decon.doublets <- gsub(\"\\\\.\",\"-\",decon.doublets)\n",
    "\n",
    "# DoubletFinder\n",
    "# Add modfieid Parameter sweep function to regress out nCOunt RNA if needed\n",
    "# Add if else statement to regress out nCount RNA if needed before running DoubletFinder\n",
    "#########################################################################################\n",
    "library(DoubletFinder)\n",
    "# Estimate Doublets\n",
    "#################################################################################################\n",
    "## pK Identification (no ground-truth) ---------------------------------------------------------------------------------------\n",
    "sweep.res.list <- paramSweep_v3(rna, PCs = 1:50, sct = FALSE)\n",
    "\n",
    "sweep.stats <- summarizeSweep(sweep.res.list , GT = FALSE)\n",
    "\n",
    "bcmvn<- find.pK(sweep.stats)\n",
    "\n",
    "pK.1 <- as.numeric(unfactor(dplyr::arrange(bcmvn,desc(BCmetric))$pK[1]))\n",
    "\n",
    "\n",
    "nExp_poi <- round(doublet.rate*length(colnames(counts)))  ## Assuming 4.6% doublet formation rate - tailor for your dataset\n",
    "\n",
    "# Run doubletFinder with pk.1\n",
    "rna <- doubletFinder_v3(rna, PCs = 1:50, pN = 0.25, pK = pK.1, nExp = nExp_poi, reuse.pANN = FALSE, sct = FALSE)\n",
    "meta <- rna@meta.data\n",
    "\n",
    "\n",
    "doublet.column <- paste0(\"DF.classifications_0.25_\",pK.1,\"_\",nExp_poi)\n",
    "\n",
    "\n",
    "doublet.calls <- rna[[doublet.column]]\n",
    "colnames(doublet.calls) <- \"Call\"\n",
    "\n",
    "rna.dub <- dplyr::filter(doublet.calls, Call == \"Doublet\")\n",
    "rna.singlet <- dplyr::filter(doublet.calls, Call == \"Singlet\")\n",
    "\n",
    "DF.doublets <- rownames(rna.dub)\n",
    "\n",
    "\n",
    "# Intersect doublet calls\n",
    "###########################################################################\n",
    "\n",
    "head(DF.doublets)\n",
    "head(decon.doublets)\n",
    "\n",
    "doublets <- intersect(decon.doublets,DF.doublets)\n",
    "rna@meta.data$Doublet.Call <- ifelse(rownames(rna@meta.data) %in% doublets,\"TRUE\",\"FALSE\")\n",
    "FeatureScatter(rna,feature1 = \"nCount_RNA\",feature2 = \"nFeature_RNA\",group.by = \"Doublet.Call\")+ggsave(\"Doublet_calls_FeatureScatter.png\")\n",
    "DimPlot(rna,group.by = \"Doublet.Call\")+ggsave(\"Doublet_calls_UMAP.png\")\n",
    "saveRDS(doublets,\"Intersection_DoubletDecon_DoubletFinder_doublets.rds\")\n",
    "cells <- colnames(rna)\n",
    "##################################################################################################################\n",
    "\n",
    "###########################################################\n",
    "# Part 3: scRNA-seq processing after doublet removal\n",
    "###########################################################\n",
    "setwd(dir)\n",
    "\n",
    "# Load the RNA dataset\n",
    "counts.init <- Read10X_h5(filename = \"./filtered_feature_bc_matrix.h5\")\n",
    "# Initialize the Seurat object with the raw (non-normalized data).\n",
    "rna <- CreateSeuratObject(counts = counts.init, min.cells = 3)# genes not present in at least 3 cells are removed\n",
    "rna\n",
    "\n",
    "rna <- rna[,cells]\n",
    "rna <- rna[,!(colnames(rna) %in% doublets)]\n",
    "dim(rna)\n",
    "\n",
    "\n",
    "#Normalize\n",
    "rna <- NormalizeData(rna, normalization.method = \"LogNormalize\", scale.factor = 10000)\n",
    "\n",
    "#Feature Selection\n",
    "rna <- FindVariableFeatures(rna, selection.method = \"vst\", nfeatures = 2000)\n",
    "\n",
    "#Scaling\n",
    "all.genes <- rownames(rna)\n",
    "rna <- ScaleData(rna, features = all.genes)\n",
    "rna <- RunPCA(rna)\n",
    "\n",
    "\n",
    "feats <- list(stromal,immune,fibroblast,endothelial,epithelial,smooth,plasma)\n",
    "\n",
    "rna <- AddModuleScore(rna,features = feats,name = c(\"stromal.\",\"immune.\",\"fibroblast.\",\"endothelial.\",\n",
    "                                                    \"epithelial.\",\"smooth.\",\"plasma.\"),search = T)\n",
    "\n",
    "\n",
    "# Add PC1 to metadata\n",
    "rna@meta.data$PC1 <- rna@reductions$pca@cell.embeddings[,1]\n",
    "\n",
    "count_cor_PC1 <- cor(rna$PC1,rna$nCount_RNA,method = \"spearman\")\n",
    "\n",
    "stromal.cor <- cor(rna$PC1,rna$stromal.1,method = \"spearman\")\n",
    "immune.cor <- cor(rna$PC1,rna$immune.2,method = \"spearman\")\n",
    "fibroblast.cor <- cor(rna$PC1,rna$fibroblast.3,method = \"spearman\")\n",
    "endothelial.cor <- cor(rna$PC1,rna$endothelial.4,method = \"spearman\")\n",
    "epithelial.cor <- cor(rna$PC1,rna$epithelial.5,method = \"spearman\")\n",
    "smooth.cor <- cor(rna$PC1,rna$smooth.6,method = \"spearman\")\n",
    "plasma.cor <- cor(rna$PC1,rna$plasma.7,method = \"spearman\")\n",
    "\n",
    "\n",
    "# Make JackStraw Plot\n",
    "rna <- JackStraw(rna, num.replicate = 100,dims = 50)\n",
    "rna <- ScoreJackStraw(rna,dims = 1:50)\n",
    "JackStrawPlot(rna, dims = 1:50)+ggsave(\"JackStraw_postdoublet.png\")\n",
    "\n",
    "\n",
    "# If PC1 is correalted with read depth, check to see if biological variation is corralted to PC1\n",
    "if (round(abs(count_cor_PC1),2) > 0.5){\n",
    "  \n",
    "  if( round(abs(stromal.cor),2) >= 0.5 |\n",
    "      round(abs(immune.cor),2) >= 0.5 |\n",
    "      round(abs(fibroblast.cor),2) >= 0.5 |\n",
    "      round(abs(endothelial.cor),2) >= 0.5 |\n",
    "      round(abs(epithelial.cor),2) >= 0.5 |\n",
    "      round(abs(smooth.cor),2) >= 0.5 |\n",
    "      round(abs(plasma.cor),2) >= 0.5){\n",
    "    \n",
    "    rna <- FindNeighbors(rna,dims = 1:50)\n",
    "    rna <- FindClusters(rna,resolution = 0.7)\n",
    "    rna <- RunUMAP(rna,dims = 1:50)\n",
    "    Idents(rna) <- \"RNA_snn_res.0.7\"\n",
    "    \n",
    "    \n",
    "    # Verify with inferCNV: is PC1 correlated with CNV events/Malignancy?\n",
    "    #########################################################################\n",
    "    # inferCNV: does PC1 also correlated with CNV/malignancy status?\n",
    "    library(infercnv)\n",
    "    library(stringr)\n",
    "    library(Seurat)\n",
    "    counts_matrix = GetAssayData(rna, slot=\"counts\")\n",
    "    \n",
    "    # Identify immune clusters\n",
    "    #######################################################\n",
    "    # Find immune cells by relative enrichment of ESTIMATE immune signature\n",
    "    library(psych)\n",
    "    test <- VlnPlot(rna,features = \"immune.2\")\n",
    "    data <- describeBy(test$data$immune.2, test$data$ident, mat = TRUE)\n",
    "    data.immune <- dplyr::filter(data,median > 0.1)\n",
    "    \n",
    "    test <- VlnPlot(rna,features = \"plasma.7\")\n",
    "    data <- describeBy(test$data$plasma.7, test$data$ident, mat = TRUE)\n",
    "    data.plasma <- dplyr::filter(data,median > 0.1)\n",
    "    \n",
    "    immune.clusters <- intersect(data.immune$group1,levels(Idents(rna)))\n",
    "    plasma.clusters <- intersect(data.plasma$group1,levels(Idents(rna)))\n",
    "    \n",
    "    immune.clusters <- unique(append(immune.clusters,plasma.clusters))\n",
    "    \n",
    "    for (i in 1:length(immune.clusters)){\n",
    "      j <- which(levels(Idents(rna)) == immune.clusters[i])\n",
    "      levels(Idents(rna))[j] <- paste0(\"immune.\",immune.clusters[i])\n",
    "    }\n",
    "    rna@meta.data$postdoublet.idents <- Idents(rna)\n",
    "    idents <- data.frame(rownames(rna@meta.data),rna@meta.data$postdoublet.idents)\n",
    "    \n",
    "    \n",
    "    colnames(idents) <- c(\"V1\",\"V2\")\n",
    "    saveRDS(rna,\"./rna_postdoublet_preinferCNV.rds\")\n",
    "    # Make inferCNV inputs\n",
    "    \n",
    "    rownames(idents) <- NULL\n",
    "    colnames(idents) <- NULL\n",
    "    write.table(idents,\"./sample_annotation_file_inferCNV.txt\",sep = \"\\t\",row.names = FALSE)\n",
    "    idents <- read.delim(\"./sample_annotation_file_inferCNV.txt\",header = F)\n",
    "    \n",
    "    \n",
    "    \n",
    "    gtf <- read.delim(GRCH38.annotations,header = F)\n",
    "    library(EnsDb.Hsapiens.v86)\n",
    "    convert.symbol = function(Data){\n",
    "      ensembls <- Data$V1\n",
    "      ensembls <- gsub(\"\\\\.[0-9]*$\", \"\", ensembls)\n",
    "      geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v86, keys= ensembls, keytype = \"GENEID\", columns = \"SYMBOL\")\n",
    "      Data <- cbind.fill(Data, geneIDs1, fill = NA)\n",
    "      Data <- na.omit(Data)\n",
    "      Data$feature <- Data$SYMBOL\n",
    "      Data.new <- data.frame(Data$SYMBOL,Data$V2,Data$V3,Data$V4)\n",
    "      Data.new$Data.V2 <- paste(\"chr\",Data.new$Data.V2,sep = \"\")\n",
    "      Data.new$Data.SYMBOL <- make.unique(Data.new$Data.SYMBOL)\n",
    "      return(Data.new)\n",
    "    }\n",
    "    gtf <- convert.symbol(gtf)\n",
    "    head(gtf)\n",
    "    \n",
    "    write.table(gtf,\"./Homo_sapiens.GRCh38.86.symbol.txt\",sep = \"\\t\",row.names = FALSE,col.names = FALSE)\n",
    "    \n",
    "    \n",
    "    num.immune.clusters = length(immune.clusters)\n",
    "    # create the infercnv object\n",
    "    if ( num.immune.clusters == 1) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=paste0(\"immune.\",immune.clusters[1]))\n",
    "      \n",
    "    } else if (num.immune.clusters == 2) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2])))\n",
    "      \n",
    "    } else if ( num.immune.clusters == 3 ) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3])))\n",
    "    } else if (num.immune.clusters == 4) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4])))\n",
    "    } else if (num.immune.clusters == 5) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5])))\n",
    "    } else if (num.immune.clusters == 6) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6])))\n",
    "    }else if (num.immune.clusters == 7) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7])))\n",
    "    }else if (num.immune.clusters == 8) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8])))\n",
    "    }else if (num.immune.clusters == 9) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[9])))\n",
    "    }else if (num.immune.clusters == 10) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[9]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[10])))\n",
    "    }\n",
    "    # perform infercnv operations to reveal cnv signal\n",
    "    infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                                 cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                                 out_dir=\"./output_dir_CNV_postdoublet_PassedPC1Checks\",  # dir is auto-created for storing outputs\n",
    "                                 cluster_by_groups=T,   # cluster\n",
    "                                 denoise=T,scale_data = T,\n",
    "                                 HMM=T,HMM_type = \"i6\",analysis_mode = \"samples\",min_cells_per_gene = 10,\n",
    "                                 BayesMaxPNormal = 0.4, num_threads = 8\n",
    "                                 \n",
    "    )\n",
    "    \n",
    "    \n",
    "    \n",
    "    regions <- read.delim(\"./output_dir_CNV_postdoublet_PassedPC1Checks/HMM_CNV_predictions.HMMi6.hmm_mode-samples.Pnorm_0.4.pred_cnv_regions.dat\")\n",
    "    probs <- read.delim(\"./output_dir_CNV_postdoublet_PassedPC1Checks/BayesNetOutput.HMMi6.hmm_mode-samples/CNV_State_Probabilities.dat\")\n",
    "    \n",
    "    probs <- as.data.frame(t(probs[3,]))\n",
    "    colnames(probs) <- \"Prob.Normal\"\n",
    "    probs <- dplyr::filter(probs,Prob.Normal < 0.05)\n",
    "    cnvs <- rownames(probs)\n",
    "    cnvs <- gsub(\"\\\\.\",\"-\",cnvs)\n",
    "    \n",
    "    regions <- regions[regions$cnv_name %in% cnvs, ]\n",
    "    \n",
    "    \n",
    "    cnv.groups <- sub(\"\\\\..*\", \"\", regions$cell_group_name)\n",
    "    \n",
    "    \n",
    "    length(which(rownames(rna@reductions$pca@cell.embeddings) == rownames(rna@meta.data)))\n",
    "    rna$PC1.loading <- rna@reductions$pca@cell.embeddings[,1]\n",
    "    rna$cell.barcode <- rownames(rna@meta.data)\n",
    "    rna$CNV.Pos <- ifelse(as.character(rna$postdoublet.idents) %in% cnv.groups,1,0)\n",
    "    \n",
    "    \n",
    "    cnv.freq <- data.frame(table(regions$cell_group_name))\n",
    "    cnv.freq$Var1 <- sub(\"\\\\..*\", \"\", cnv.freq$Var1)\n",
    "    \n",
    "    rna$Total_CNVs <- ifelse(as.character(rna$postdoublet.idents) %in% cnv.freq$Var1,cnv.freq$Freq,0)\n",
    "    \n",
    "    boxplot.cnv <- ggplot(rna@meta.data,aes(x= postdoublet.idents,y=PC1.loading,color = as.factor(CNV.Pos)))+geom_boxplot()\n",
    "    boxplot.cnv+ggsave(\"postdoublet_CNV_PC1_boxplot.png\")\n",
    "    \n",
    "    data <- describeBy(boxplot.cnv$data$PC1.loading, boxplot.cnv$data$postdoublet.idents, mat = TRUE)\n",
    "    data$CNV <- ifelse(data$group1 %in% cnv.groups,1,0)\n",
    "    \n",
    "    wilcox <- wilcox.test(data = rna@meta.data,PC1.loading~CNV.Pos)\n",
    "    \n",
    "    if (wilcox$p.value < 0.05){\n",
    "      rna <- rna\n",
    "      library(stringr)\n",
    "      levels(Idents(rna)) <- str_remove(levels(Idents(rna)),\"immune.\")\n",
    "      saveRDS(rna,\"./rna_postdoublet_PassedPC1Checks.rds\")\n",
    "    }else{\n",
    "      all.genes <- rownames(rna)\n",
    "      rna <- ScaleData(rna, features = all.genes,vars.to.regress = \"nCount_RNA\")\n",
    "      rna <- FindNeighbors(rna,dims = 1:50)\n",
    "      rna <- FindClusters(rna,resolution = 0.7)\n",
    "      rna <- RunUMAP(rna,dims = 1:50)\n",
    "      Idents(rna) <- \"RNA_snn_res.0.7\"\n",
    "      \n",
    "      library(infercnv)\n",
    "      library(stringr)\n",
    "      library(Seurat)\n",
    "      counts_matrix = GetAssayData(rna, slot=\"counts\")\n",
    "      \n",
    "      \n",
    "      # Identify immune clusters\n",
    "      #######################################################\n",
    "      # Find immune cells by relative enrichment of ESTIMATE immune signature\n",
    "      library(psych)\n",
    "      test <- VlnPlot(rna,features = \"immune.2\")\n",
    "      data <- describeBy(test$data$immune.2, test$data$ident, mat = TRUE)\n",
    "      data.immune <- dplyr::filter(data,median > 0.1)\n",
    "      \n",
    "      test <- VlnPlot(rna,features = \"plasma.7\")\n",
    "      data <- describeBy(test$data$plasma.7, test$data$ident, mat = TRUE)\n",
    "      data.plasma <- dplyr::filter(data,median > 0.1)\n",
    "      \n",
    "      immune.clusters <- intersect(data.immune$group1,levels(Idents(rna)))\n",
    "      plasma.clusters <- intersect(data.plasma$group1,levels(Idents(rna)))\n",
    "      \n",
    "      immune.clusters <- unique(append(immune.clusters,plasma.clusters))\n",
    "      \n",
    "      for (i in 1:length(immune.clusters)){\n",
    "        j <- which(levels(Idents(rna)) == immune.clusters[i])\n",
    "        levels(Idents(rna))[j] <- paste0(\"immune.\",immune.clusters[i])\n",
    "      }\n",
    "      rna@meta.data$postdoublet.idents <- Idents(rna)\n",
    "      idents <- data.frame(rownames(rna@meta.data),rna@meta.data$postdoublet.idents)\n",
    "      saveRDS(rna,\"./rna_postdoublet_preinferCNV.rds\")\n",
    "      #MAke inferCNV input\n",
    "      \n",
    "      colnames(idents) <- c(\"V1\",\"V2\")\n",
    "      \n",
    "      \n",
    "      rownames(idents) <- NULL\n",
    "      colnames(idents) <- NULL\n",
    "      write.table(idents,\"./sample_annotation_file_inferCNV.txt\",sep = \"\\t\",row.names = FALSE)\n",
    "      idents <- read.delim(\"./sample_annotation_file_inferCNV.txt\",header = F)\n",
    "      \n",
    "      \n",
    "      \n",
    "      gtf <- read.delim(GRCH38.annotations,header = F)\n",
    "      library(EnsDb.Hsapiens.v86)\n",
    "      convert.symbol = function(Data){\n",
    "        ensembls <- Data$V1\n",
    "        ensembls <- gsub(\"\\\\.[0-9]*$\", \"\", ensembls)\n",
    "        geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v86, keys= ensembls, keytype = \"GENEID\", columns = \"SYMBOL\")\n",
    "        Data <- cbind.fill(Data, geneIDs1, fill = NA)\n",
    "        Data <- na.omit(Data)\n",
    "        Data$feature <- Data$SYMBOL\n",
    "        Data.new <- data.frame(Data$SYMBOL,Data$V2,Data$V3,Data$V4)\n",
    "        Data.new$Data.V2 <- paste(\"chr\",Data.new$Data.V2,sep = \"\")\n",
    "        Data.new$Data.SYMBOL <- make.unique(Data.new$Data.SYMBOL)\n",
    "        return(Data.new)\n",
    "      }\n",
    "      gtf <- convert.symbol(gtf)\n",
    "      head(gtf)\n",
    "      \n",
    "      write.table(gtf,\"./Homo_sapiens.GRCh38.86.symbol.txt\",sep = \"\\t\",row.names = FALSE,col.names = FALSE)\n",
    "      \n",
    "      \n",
    "      num.immune.clusters = length(immune.clusters)\n",
    "      # create the infercnv object\n",
    "      if ( num.immune.clusters == 1) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=paste0(\"immune.\",immune.clusters[1]))\n",
    "        \n",
    "      } else if (num.immune.clusters == 2) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2])))\n",
    "        \n",
    "      } else if ( num.immune.clusters == 3 ) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[3])))\n",
    "      } else if (num.immune.clusters == 4) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[4])))\n",
    "      } else if (num.immune.clusters == 5) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[5])))\n",
    "      } else if (num.immune.clusters == 6) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[6])))\n",
    "      }else if (num.immune.clusters == 7) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[7])))\n",
    "      }else if (num.immune.clusters == 8) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[8])))\n",
    "      }else if (num.immune.clusters == 9) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[9])))\n",
    "      }else if (num.immune.clusters == 10) {\n",
    "        infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                            annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                            gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                            ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[9]),\n",
    "                                                              paste0(\"immune.\",immune.clusters[10])))\n",
    "      }\n",
    "      # perform infercnv operations to reveal cnv signal\n",
    "      infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                                   cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                                   out_dir=\"./output_dir_CNV_postdoublet_FailedCNVTest\",  # dir is auto-created for storing outputs\n",
    "                                   cluster_by_groups=T,   # cluster\n",
    "                                   denoise=T,scale_data = T,\n",
    "                                   HMM=T,HMM_type = \"i6\",analysis_mode = \"samples\",min_cells_per_gene = 10,\n",
    "                                   BayesMaxPNormal = 0.4, num_threads = 8\n",
    "                                   \n",
    "      )\n",
    "      \n",
    "      \n",
    "      \n",
    "      regions <- read.delim(\"./output_dir_CNV_postdoublet_FailedCNVTest/HMM_CNV_predictions.HMMi6.hmm_mode-samples.Pnorm_0.4.pred_cnv_regions.dat\")\n",
    "      probs <- read.delim(\"./output_dir_CNV_postdoublet_FailedCNVTest/BayesNetOutput.HMMi6.hmm_mode-samples/CNV_State_Probabilities.dat\")\n",
    "      \n",
    "      probs <- as.data.frame(t(probs[3,]))\n",
    "      colnames(probs) <- \"Prob.Normal\"\n",
    "      probs <- dplyr::filter(probs,Prob.Normal < 0.05)\n",
    "      cnvs <- rownames(probs)\n",
    "      cnvs <- gsub(\"\\\\.\",\"-\",cnvs)\n",
    "      \n",
    "      regions <- regions[regions$cnv_name %in% cnvs, ]\n",
    "      \n",
    "      \n",
    "      cnv.groups <- sub(\"\\\\..*\", \"\", regions$cell_group_name)\n",
    "      \n",
    "      \n",
    "      length(which(rownames(rna@reductions$pca@cell.embeddings) == rownames(rna@meta.data)))\n",
    "      rna$PC1.loading <- rna@reductions$pca@cell.embeddings[,1]\n",
    "      rna$cell.barcode <- rownames(rna@meta.data)\n",
    "      rna$CNV.Pos <- ifelse(as.character(rna$postdoublet.idents) %in% cnv.groups,1,0)\n",
    "      \n",
    "      \n",
    "      cnv.freq <- data.frame(table(regions$cell_group_name))\n",
    "      cnv.freq$Var1 <- sub(\"\\\\..*\", \"\", cnv.freq$Var1)\n",
    "      \n",
    "      rna$Total_CNVs <- ifelse(as.character(rna$postdoublet.idents) %in% cnv.freq$Var1,cnv.freq$Freq,0)\n",
    "      \n",
    "      saveRDS(rna,\"./rna_postdoublet_FailedCNVTest.rds\")\n",
    "    }\n",
    "\n",
    "  }else{\n",
    "    all.genes <- rownames(rna)\n",
    "    rna <- ScaleData(rna, features = all.genes,vars.to.regress = \"nCount_RNA\")\n",
    "    rna <- FindNeighbors(rna,dims = 1:50)\n",
    "    rna <- FindClusters(rna,resolution = 0.7)\n",
    "    rna <- RunUMAP(rna,dims = 1:50)\n",
    "    Idents(rna) <- \"RNA_snn_res.0.7\"\n",
    "    \n",
    "    # Proceed with CNV\n",
    "    library(infercnv)\n",
    "    library(stringr)\n",
    "    library(Seurat)\n",
    "    counts_matrix = GetAssayData(rna, slot=\"counts\")\n",
    "    \n",
    "    # Identify immune clusters\n",
    "    #######################################################\n",
    "    # Find immune cells by relative enrichment of ESTIMATE immune signature\n",
    "    library(psych)\n",
    "    test <- VlnPlot(rna,features = \"immune.2\")\n",
    "    data <- describeBy(test$data$immune.2, test$data$ident, mat = TRUE)\n",
    "    data.immune <- dplyr::filter(data,median > 0.1)\n",
    "    \n",
    "    test <- VlnPlot(rna,features = \"plasma.7\")\n",
    "    data <- describeBy(test$data$plasma.7, test$data$ident, mat = TRUE)\n",
    "    data.plasma <- dplyr::filter(data,median > 0.1)\n",
    "    \n",
    "    immune.clusters <- intersect(data.immune$group1,levels(Idents(rna)))\n",
    "    plasma.clusters <- intersect(data.plasma$group1,levels(Idents(rna)))\n",
    "    \n",
    "    immune.clusters <- unique(append(immune.clusters,plasma.clusters))\n",
    "    \n",
    "    for (i in 1:length(immune.clusters)){\n",
    "      j <- which(levels(Idents(rna)) == immune.clusters[i])\n",
    "      levels(Idents(rna))[j] <- paste0(\"immune.\",immune.clusters[i])\n",
    "    }\n",
    "    rna@meta.data$postdoublet.idents <- Idents(rna)\n",
    "    idents <- data.frame(rownames(rna@meta.data),rna@meta.data$postdoublet.idents)\n",
    "    \n",
    "    \n",
    "    \n",
    "    colnames(idents) <- c(\"V1\",\"V2\")\n",
    "    \n",
    "    saveRDS(rna,\"./rna_postdoublet_preinferCNV.rds\")\n",
    "    # Make inferCNV inputs\n",
    "    \n",
    "    rownames(idents) <- NULL\n",
    "    colnames(idents) <- NULL\n",
    "    write.table(idents,\"./sample_annotation_file_inferCNV.txt\",sep = \"\\t\",row.names = FALSE)\n",
    "    idents <- read.delim(\"./sample_annotation_file_inferCNV.txt\",header = F)\n",
    "    \n",
    "    \n",
    "    \n",
    "    gtf <- read.delim(GRCH38.annotations,header = F)\n",
    "    library(EnsDb.Hsapiens.v86)\n",
    "    convert.symbol = function(Data){\n",
    "      ensembls <- Data$V1\n",
    "      ensembls <- gsub(\"\\\\.[0-9]*$\", \"\", ensembls)\n",
    "      geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v86, keys= ensembls, keytype = \"GENEID\", columns = \"SYMBOL\")\n",
    "      Data <- cbind.fill(Data, geneIDs1, fill = NA)\n",
    "      Data <- na.omit(Data)\n",
    "      Data$feature <- Data$SYMBOL\n",
    "      Data.new <- data.frame(Data$SYMBOL,Data$V2,Data$V3,Data$V4)\n",
    "      Data.new$Data.V2 <- paste(\"chr\",Data.new$Data.V2,sep = \"\")\n",
    "      Data.new$Data.SYMBOL <- make.unique(Data.new$Data.SYMBOL)\n",
    "      return(Data.new)\n",
    "    }\n",
    "    gtf <- convert.symbol(gtf)\n",
    "    head(gtf)\n",
    "    \n",
    "    write.table(gtf,\"./Homo_sapiens.GRCh38.86.symbol.txt\",sep = \"\\t\",row.names = FALSE,col.names = FALSE)\n",
    "    \n",
    "    \n",
    "    num.immune.clusters = length(immune.clusters)\n",
    "    # create the infercnv object\n",
    "    if ( num.immune.clusters == 1) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=paste0(\"immune.\",immune.clusters[1]))\n",
    "      \n",
    "    } else if (num.immune.clusters == 2) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2])))\n",
    "      \n",
    "    } else if ( num.immune.clusters == 3 ) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3])))\n",
    "    } else if (num.immune.clusters == 4) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4])))\n",
    "    } else if (num.immune.clusters == 5) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5])))\n",
    "    } else if (num.immune.clusters == 6) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6])))\n",
    "    }else if (num.immune.clusters == 7) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7])))\n",
    "    }else if (num.immune.clusters == 8) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8])))\n",
    "    }else if (num.immune.clusters == 9) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[9])))\n",
    "    }else if (num.immune.clusters == 10) {\n",
    "      infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                          annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                          gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                          ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[9]),\n",
    "                                                            paste0(\"immune.\",immune.clusters[10])))\n",
    "    }\n",
    "    # perform infercnv operations to reveal cnv signal\n",
    "    infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                                 cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                                 out_dir=\"./output_dir_CNV_postdoublet_FailedCorTest\",  # dir is auto-created for storing outputs\n",
    "                                 cluster_by_groups=T,   # cluster\n",
    "                                 denoise=T,scale_data = T,\n",
    "                                 HMM=T,HMM_type = \"i6\",analysis_mode = \"samples\",min_cells_per_gene = 10,\n",
    "                                 BayesMaxPNormal = 0.4, num_threads = 8\n",
    "                                 \n",
    "    )\n",
    "    \n",
    "    \n",
    "    \n",
    "    regions <- read.delim(\"./output_dir_CNV_postdoublet_FailedCorTest/HMM_CNV_predictions.HMMi6.hmm_mode-samples.Pnorm_0.4.pred_cnv_regions.dat\")\n",
    "    probs <- read.delim(\"./output_dir_CNV_postdoublet_FailedCorTest/BayesNetOutput.HMMi6.hmm_mode-samples/CNV_State_Probabilities.dat\")\n",
    "    \n",
    "    probs <- as.data.frame(t(probs[3,]))\n",
    "    colnames(probs) <- \"Prob.Normal\"\n",
    "    probs <- dplyr::filter(probs,Prob.Normal < 0.05)\n",
    "    cnvs <- rownames(probs)\n",
    "    cnvs <- gsub(\"\\\\.\",\"-\",cnvs)\n",
    "    \n",
    "    regions <- regions[regions$cnv_name %in% cnvs, ]\n",
    "    \n",
    "    \n",
    "    cnv.groups <- sub(\"\\\\..*\", \"\", regions$cell_group_name)\n",
    "    \n",
    "    \n",
    "    length(which(rownames(rna@reductions$pca@cell.embeddings) == rownames(rna@meta.data)))\n",
    "    rna$PC1.loading <- rna@reductions$pca@cell.embeddings[,1]\n",
    "    rna$cell.barcode <- rownames(rna@meta.data)\n",
    "    rna$CNV.Pos <- ifelse(as.character(rna$postdoublet.idents) %in% cnv.groups,1,0)\n",
    "    \n",
    "    \n",
    "    cnv.freq <- data.frame(table(regions$cell_group_name))\n",
    "    cnv.freq$Var1 <- sub(\"\\\\..*\", \"\", cnv.freq$Var1)\n",
    "    \n",
    "    rna$Total_CNVs <- ifelse(as.character(rna$postdoublet.idents) %in% cnv.freq$Var1,cnv.freq$Freq,0)\n",
    "    saveRDS(rna,\"./rna_postdoublet_FailedCorTest.rds\")\n",
    "  }\n",
    "}else{\n",
    "  \n",
    "  rna <- FindNeighbors(rna,dims = 1:50)\n",
    "  rna <- FindClusters(rna,resolution = 0.7)\n",
    "  rna <- RunUMAP(rna,dims = 1:50)\n",
    "  Idents(rna) <- \"RNA_snn_res.0.7\"\n",
    "  \n",
    "  # Proceed with CNV\n",
    "  library(infercnv)\n",
    "  library(stringr)\n",
    "  library(Seurat)\n",
    "  counts_matrix = GetAssayData(rna, slot=\"counts\")\n",
    "  \n",
    "  \n",
    "  # Identify immune clusters\n",
    "  #######################################################\n",
    "  # Find immune cells by relative enrichment of ESTIMATE immune signature\n",
    "  library(psych)\n",
    "  test <- VlnPlot(rna,features = \"immune.2\")\n",
    "  data <- describeBy(test$data$immune.2, test$data$ident, mat = TRUE)\n",
    "  data.immune <- dplyr::filter(data,median > 0.1)\n",
    "  \n",
    "  test <- VlnPlot(rna,features = \"plasma.7\")\n",
    "  data <- describeBy(test$data$plasma.7, test$data$ident, mat = TRUE)\n",
    "  data.plasma <- dplyr::filter(data,median > 0.1)\n",
    "  \n",
    "  immune.clusters <- intersect(data.immune$group1,levels(Idents(rna)))\n",
    "  plasma.clusters <- intersect(data.plasma$group1,levels(Idents(rna)))\n",
    "  \n",
    "  immune.clusters <- unique(append(immune.clusters,plasma.clusters))\n",
    "  \n",
    "  for (i in 1:length(immune.clusters)){\n",
    "    j <- which(levels(Idents(rna)) == immune.clusters[i])\n",
    "    levels(Idents(rna))[j] <- paste0(\"immune.\",immune.clusters[i])\n",
    "  }\n",
    "  rna@meta.data$postdoublet.idents <- Idents(rna)\n",
    "  idents <- data.frame(rownames(rna@meta.data),rna@meta.data$postdoublet.idents)\n",
    "  \n",
    "  \n",
    "  colnames(idents) <- c(\"V1\",\"V2\")\n",
    "  saveRDS(rna,\"./rna_postdoublet_preinferCNV.rds\")\n",
    "  # Make inferCNV inputs\n",
    "  rownames(idents) <- NULL\n",
    "  colnames(idents) <- NULL\n",
    "  write.table(idents,\"./sample_annotation_file_inferCNV.txt\",sep = \"\\t\",row.names = FALSE)\n",
    "  idents <- read.delim(\"./sample_annotation_file_inferCNV.txt\",header = F)\n",
    "  \n",
    "  \n",
    "  \n",
    "  gtf <- read.delim(GRCH38.annotations,header = F)\n",
    "  library(EnsDb.Hsapiens.v86)\n",
    "  convert.symbol = function(Data){\n",
    "    ensembls <- Data$V1\n",
    "    ensembls <- gsub(\"\\\\.[0-9]*$\", \"\", ensembls)\n",
    "    geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v86, keys= ensembls, keytype = \"GENEID\", columns = \"SYMBOL\")\n",
    "    Data <- cbind.fill(Data, geneIDs1, fill = NA)\n",
    "    Data <- na.omit(Data)\n",
    "    Data$feature <- Data$SYMBOL\n",
    "    Data.new <- data.frame(Data$SYMBOL,Data$V2,Data$V3,Data$V4)\n",
    "    Data.new$Data.V2 <- paste(\"chr\",Data.new$Data.V2,sep = \"\")\n",
    "    Data.new$Data.SYMBOL <- make.unique(Data.new$Data.SYMBOL)\n",
    "    return(Data.new)\n",
    "  }\n",
    "  gtf <- convert.symbol(gtf)\n",
    "  head(gtf)\n",
    "  \n",
    "  write.table(gtf,\"./Homo_sapiens.GRCh38.86.symbol.txt\",sep = \"\\t\",row.names = FALSE,col.names = FALSE)\n",
    "  \n",
    "  \n",
    "  num.immune.clusters = length(immune.clusters)\n",
    "  # create the infercnv object\n",
    "  # create the infercnv object\n",
    "  if ( num.immune.clusters == 1) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=paste0(\"immune.\",immune.clusters[1]))\n",
    "    \n",
    "  } else if (num.immune.clusters == 2) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2])))\n",
    "    \n",
    "  } else if ( num.immune.clusters == 3 ) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[3])))\n",
    "  } else if (num.immune.clusters == 4) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[4])))\n",
    "  } else if (num.immune.clusters == 5) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[5])))\n",
    "  } else if (num.immune.clusters == 6) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[6])))\n",
    "  }else if (num.immune.clusters == 7) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[7])))\n",
    "  }else if (num.immune.clusters == 8) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[8])))\n",
    "  }else if (num.immune.clusters == 9) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[9])))\n",
    "  }else if (num.immune.clusters == 10) {\n",
    "    infercnv_obj = CreateInfercnvObject(raw_counts_matrix=as.matrix(counts_matrix),\n",
    "                                        annotations_file=\"./sample_annotation_file_inferCNV.txt\",\n",
    "                                        gene_order_file=\"./Homo_sapiens.GRCh38.86.symbol.txt\",\n",
    "                                        ref_group_names=c(paste0(\"immune.\",immune.clusters[1]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[2]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[3]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[4]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[5]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[6]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[7]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[8]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[9]),\n",
    "                                                          paste0(\"immune.\",immune.clusters[10])))\n",
    "  }\n",
    "  # perform infercnv operations to reveal cnv signal\n",
    "  infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                               cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                               out_dir=\"./output_dir_CNV_postdoublet_SkipChecks\",  # dir is auto-created for storing outputs\n",
    "                               cluster_by_groups=T,   # cluster\n",
    "                               denoise=T,scale_data = T,\n",
    "                               HMM=T,HMM_type = \"i6\",analysis_mode = \"samples\",min_cells_per_gene = 10,\n",
    "                               BayesMaxPNormal = 0.4, num_threads = 8\n",
    "                               \n",
    "  )\n",
    "  \n",
    "  \n",
    "  \n",
    "  regions <- read.delim(\"./output_dir_CNV_postdoublet_SkipChecks/HMM_CNV_predictions.HMMi6.hmm_mode-samples.Pnorm_0.4.pred_cnv_regions.dat\")\n",
    "  probs <- read.delim(\"./output_dir_CNV_postdoublet_SkipChecks/BayesNetOutput.HMMi6.hmm_mode-samples/CNV_State_Probabilities.dat\")\n",
    "  \n",
    "  probs <- as.data.frame(t(probs[3,]))\n",
    "  colnames(probs) <- \"Prob.Normal\"\n",
    "  probs <- dplyr::filter(probs,Prob.Normal < 0.05)\n",
    "  cnvs <- rownames(probs)\n",
    "  cnvs <- gsub(\"\\\\.\",\"-\",cnvs)\n",
    "  \n",
    "  regions <- regions[regions$cnv_name %in% cnvs, ]\n",
    "  \n",
    "  \n",
    "  cnv.groups <- sub(\"\\\\..*\", \"\", regions$cell_group_name)\n",
    "  \n",
    "  \n",
    "  length(which(rownames(rna@reductions$pca@cell.embeddings) == rownames(rna@meta.data)))\n",
    "  rna$PC1.loading <- rna@reductions$pca@cell.embeddings[,1]\n",
    "  rna$cell.barcode <- rownames(rna@meta.data)\n",
    "  rna$CNV.Pos <- ifelse(as.character(rna$postdoublet.idents) %in% cnv.groups,1,0)\n",
    "  \n",
    "  \n",
    "  cnv.freq <- data.frame(table(regions$cell_group_name))\n",
    "  cnv.freq$Var1 <- sub(\"\\\\..*\", \"\", cnv.freq$Var1)\n",
    "  \n",
    "  rna$Total_CNVs <- ifelse(as.character(rna$postdoublet.idents) %in% cnv.freq$Var1,cnv.freq$Freq,0)\n",
    "  \n",
    "  \n",
    "  boxplot.cnv <- ggplot(rna@meta.data,aes(x= postdoublet.idents,y=PC1.loading,color = as.factor(CNV.Pos)))+geom_boxplot()\n",
    "  boxplot.cnv+ggsave(\"Postdoublet_CNV_PC1_boxplot.png\")\n",
    "  saveRDS(rna,\"./rna_postdoublet_SkipChecks.rds\")\n",
    "}\n",
    "\n",
    "Idents(rna)<- \"RNA_snn_res.0.7\"\n",
    "\n",
    "# DEG analysis with Wilcox\n",
    "##########################################################################\n",
    "\n",
    "# Wilcox\n",
    "Wilcox.markers <- FindAllMarkers(object =rna, min.pct = 0.25,only.pos = F,\n",
    "                                 test.use = \"wilcox\")\n",
    "saveRDS(Wilcox.markers,\"./wilcox_DEGs.rds\")\n",
    "\n",
    "\n",
    "###########################################################\n",
    "# Part 4: SingleR cell typing \n",
    "###########################################################\n",
    "\n",
    "# SingleR labeling of celltypes\n",
    "##########################################################################\n",
    "rna.sce <- as.SingleCellExperiment(rna)\n",
    "\n",
    "\n",
    "# Set reference paths:\n",
    "# Wang et al. GSE111976\n",
    "ref.data.counts <- readRDS(\"./GSE111976_ct_endo_10x.rds\")\n",
    "meta <- read.csv(\"./GSE111976_summary_10x_day_donor_ctype.csv\")\n",
    "rownames(meta) <- meta$X\n",
    "\n",
    "length(which(rownames(meta) == colnames(ref.data.counts)))\n",
    "ref.data.endo <- CreateSeuratObject(ref.data.counts,meta.data = meta)\n",
    "Idents(ref.data.endo) <- \"cell_type\"\n",
    "ref.data.endo <- NormalizeData(ref.data.endo)\n",
    "\n",
    "# SingleR annotation\n",
    "#####################################################\n",
    "\n",
    "# Read in reference datasets for SingleR annotation \n",
    "\n",
    "# 1) Slyper et al. Nat. Medicine 2020 scRNA-seq ovarian tumor \n",
    "ref.data.endo <- as.SingleCellExperiment(ref.data.endo)\n",
    "\n",
    "# 2) Human Primary Cell Atlas Data (microarray)\n",
    "ref.data.HPCA <- readRDS(\"./HPCA_celldex.rds\")\n",
    "# \n",
    "# # 3) BluePrint Encode (bulk RNA-seq)\n",
    "ref.data.BED <- readRDS(\"./BluePrintEncode_celldex.rds\")\n",
    "\n",
    "\n",
    "\n",
    "# 2) Single-cell level label transfer: \n",
    "predictions.HPCA.sc <- SingleR(test=rna.sce, assay.type.test=\"logcounts\", assay.type.ref=\"logcounts\",\n",
    "                               ref=ref.data.HPCA, labels=ref.data.HPCA$label.main)\n",
    "predictions.BED.sc <- SingleR(test=rna.sce, assay.type.test=\"logcounts\", assay.type.ref=\"logcounts\",\n",
    "                              ref=ref.data.BED, labels=ref.data.BED$label.main)\n",
    "# Use de.method wilcox with scRNA-seq reference b/c the reference data is more sparse\n",
    "predictions.endo.sc <- SingleR(test=rna.sce, assay.type.test=\"logcounts\", assay.type.ref=\"logcounts\",\n",
    "                               ref=ref.data.endo, labels=ref.data.endo$cell_type,de.method = \"wilcox\")\n",
    "\n",
    "\n",
    "rna$SingleR.HPCA <- predictions.HPCA.sc$pruned.labels\n",
    "rna$SingleR.BED <- predictions.BED.sc$pruned.labels\n",
    "rna$SingleR.endo <- predictions.endo.sc$pruned.labels\n",
    "\n",
    "\n",
    "# Save Seurat object \n",
    "#date <- Sys.Date()\n",
    "saveRDS(rna,paste0(SAMPLE.ID,\"_scRNA_processed.rds\"))\n",
    "\n",
    "# ###########################################################################################################\n",
    "# # Starting cells, PostQC cells, doublets, Post doublet/QC cells, Cluster #\n",
    "output.meta <- data.frame(StartingNumCells=length(colnames(counts.init)),\n",
    "                          nMADLogCounts =2,\n",
    "                          nMADLogFeatures = 2,\n",
    "                          nMADLog1pMito =2,\n",
    "                          PostQCNumCells=PostQCNumCells,\n",
    "                          ExpectedDoubletFraction=doublet.rate,\n",
    "                          ObservedDoubletFraction=length(doublets)/length(colnames(counts.init)),\n",
    "                          PostDoubletNumCells=length(colnames(rna)),\n",
    "                          NumClusters=length(levels(Idents(rna))),\n",
    "                          DoubletFinderpK = pK.1,\n",
    "                          MinNumCounts=min(rna$nCount_RNA),\n",
    "                          MaxNumCounts= max(rna$nCount_RNA),\n",
    "                          MedianNumbCounts = median(rna$nCount_RNA),\n",
    "                          MinNumFeats=min(rna$nFeature_RNA),\n",
    "                          MaxNumFeats= max(rna$nFeature_RNA),\n",
    "                          MedianNumbFeats = median(rna$nFeature_RNA),\n",
    "                          stringsAsFactors=FALSE)\n",
    "output <- as.data.frame(t(output.meta))\n",
    "colnames(output) <- SAMPLE.ID\n",
    "xlsx::write.xlsx(output, \"scRNA_pipeline_summary.xlsx\",\n",
    "                 row.names = T, col.names = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#########################################################################################################\n",
    "# END OF SCRIPT\n",
    "#########################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6db11684-199b-4a58-bf27-0524d3cc3891",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
